### DISCUSSION TITLE
How to properly test an Uploader?

### DISCUSSION THREAD

**Discussion by danielnc** [shrinerb/shrine#604] [Q&A] [answered]
I could only find on the examples, how to test metadata extraction(such as size, width, height, etc)

I couldn't find a way to properly test the validations added either by the validation plugin or the validation_helpers

I am wondering, how can I test the code from the uploader to guarantee that the max/min size, dimensions etc are in place

What I am looking for is something similar to should_matchers how can I manually test my validations?

**Comment by janko** [comment, marked as answer]
It might be easiest to test through a specific model, especially if there is model-specific code in the uploader.

```rb
photo = Photo.new
photo.image = StringIO.new("too small")
refute photo.valid?
assert_equal ["..."], photo.errors[:image]
```

But if the uploader doesn't reference the model, it should also be possible to test the attacher individually:

```rb
attacher = ImageUploader::Attacher.new
attacher.assign StringIO.new("too small")
assert_equal ["..."], attacher.errors
```

**Reply by danielnc** [reply]
@janko thanks! that worked

for those that are looking for a similar solution this is what we did


```ruby
    describe "validations" do
      let(:uploader) {
        data = attacher.data
        data["metadata"]["size"] = 1.gigabyte
        MediaUploader::Attacher.from_data(data)
      }

      before { uploader.validate }

      it "validates the size" do
        expect(uploader.errors).to include("size must not be greater than 5.0 MB" )
      end
    end
```
